### ISSUE TITLE
LightGBM cv

### ISSUE THREAD

**Issue by Sanchita-P** [microsoft/LightGBM#2502]
How does LightGBM's cv module identify what the objective function is? I understand that LightGBM's default is: Regression, however, my problem is a classification one. There is no option to specify what the objective function is. 
Additionally, in LGBMClassifier, if I specify the objective as binary, will the loss function also be automatically chosen correctly without me having to manually specify it? I do see from the documentation, that something of this sort does happen, however confirming that it is Binary log loss in this case.
Confirmation, using Sklearn's API of LightGBM, I can predict the class, however, LightGBM only gives me the probability. Correct?
Last question, I have attempted to use LightGBM's cv module with the Sklearn module in the following way : (Could you let me know if such cross-module usage is allowed?
```
x_train, x_test, y_train, y_test = train_test_split(df_merge, labels, test_size=0.25)

n_folds = 5

lgb_train = lgb.Dataset(x_train, y_train)

def objective(params, n_folds = n_folds):
    """Objective function for Gradient Boosting Machine Hyperparameter Tuning"""
    
    print(params)
#    params['num_iterations '] = int(params['num_iterations '])
    params['max_depth'] = int(params['max_depth'])
    params['num_leaves'] = int(params['num_leaves'])
    params['subsample_for_bin'] = int(params['subsample_for_bin'])
    params['min_child_samples'] = int(params['min_child_samples'])

    # Perform n_fold cross validation with hyperparameters
    # Use early stopping and evalute based on ROC AUC
    cv_results = lgb.cv(params, lgb_train, nfold = n_folds, num_boost_round = 10000, 
                        early_stopping_rounds = 100, metrics = 'auc', seed = 50)
  
    # Extract the best score
    best_score = max(cv_results['auc-mean'])
    
    # Loss must be minimized
    loss = 1 - best_score
    
    # Dictionary with information for evaluation
    return {'loss': loss, 'params': params, 'status': STATUS_OK}

space = {
#    'class_weight': hp.choice('class_weight', [None, 'balanced']),
#    'boosting_type': 'gbdt',
    'num_leaves': hp.quniform('num_leaves', 30, 150, 1),
#    'learning_rate': hp.loguniform('learning_rate', np.log(0.01), np.log(0.2)),
    'subsample_for_bin': hp.quniform('subsample_for_bin', 20000, 300000, 20000),
    'min_child_samples': hp.quniform('min_child_samples', 20, 500, 5),
    'reg_alpha': hp.uniform('reg_alpha', 0.0, 1.0),
    'reg_lambda': hp.uniform('reg_lambda', 0.0, 1.0),
    'colsample_bytree': hp.uniform('colsample_by_tree', 0.6, 1.0),
#    'num_iterations ' : hp.quniform('num_iterations ', 100, 1000, 1),
    'max_depth' : hp.quniform('max_depth', 3, 10, 1),
    'subsample' : hp.quniform('subsample', 0.5, 1, 0.05)
}

trials = Trials()
best = fmin(objective, space, algo=tpe.suggest, trials=trials, max_evals=10)

best['max_depth'] = int(best['max_depth'])
best['num_leaves'] = int(best['num_leaves'])
best['subsample_for_bin'] = int(best['subsample_for_bin'])
best['min_child_samples'] = int(best['min_child_samples'])

best_gbm = lgb.LGBMClassifier(objective = 'binary', boosting_type = 'gbdt',**best)
best_gbm.fit(x_train, y_train)

f1_score(y_test, best_gbm.predict(x_test))
roc_auc_score(y_test, best_gbm.predict(x_test))

print('Plotting feature importances...')
ax = lgb.plot_importance(best_gbm, max_num_features=15)
plt.show()

feature_imp = pd.DataFrame()
feature_imp["feature"] = list(x_train.columns)
feature_imp["Importance"] = best_gbm.feature_importances_
feature_imp.to_clipboard()
```

Thank you so much!
@StrikerRUS @guolinke @chivee @wxchan @henry0312

**Comment by StrikerRUS** [comment]
@Sanchita-P 
> How does LightGBM's cv module identify what the objective function is? I understand that LightGBM's default is: Regression, however, my problem is a classification one. There is no option to specify what the objective function is.

You can (should) specify objective function via [objective](https://lightgbm.readthedocs.io/en/latest/Parameters.html#objective) param: `lgb.cv(params={'objective': 'binary'})`. You are right, as it's mentioned in docs, default one is `regression`.

> Additionally, in LGBMClassifier, if I specify the objective as binary, will the loss function also be automatically chosen correctly without me having to manually specify it?

"Objective and "loss" are different words for the same thing.

> Confirmation, using Sklearn's API of LightGBM, I can predict the class, however, LightGBM only gives me the probability. Correct?

Yep. For standard API (not the sklearn wrapper) you should use argmax manually to predict a class label.

> Last question, I have attempted to use LightGBM's cv module with the Sklearn module in the following way : (Could you let me know if such cross-module usage is allowed?

It's encouraged to not mix two different APIs, except the case when you use sklearn's folds generator for `folds` param of `lgb.cv` function.
https://github.com/microsoft/LightGBM/blob/dc5840709909bc18df437cddf62677f7be2915a4/python-package/lightgbm/engine.py#L390-L395

**Comment by Sanchita-P** [comment]
Hi @StrikerRUS , 

Thank you for your response! I understand that Sklearn API and LighGBM cv should not be used together. Due to this, I have made the following changes to my code but am getting an extremely strange error.

The error is : train() got an unexpected keyword argument 'bagging_freq'. This is odd since bagging_freq is a legitimate parameter. My code is this: 

x_train, x_test, y_train, y_test = train_test_split(df_dummy[df_merge.columns], labels, test_size=0.25)

n_folds = 5

lgb_train = lgb.Dataset(x_train, y_train)

def objective(params, n_folds = n_folds):
    """Objective function for Gradient Boosting Machine Hyperparameter Tuning"""
    
    print(params)
#    params['num_iterations '] = int(params['num_iterations '])
    params['max_depth'] = int(params['max_depth'])
    params['num_leaves'] = int(params['num_leaves'])
#    params['subsample_for_bin'] = int(params['subsample_for_bin'])
    params['min_child_samples'] = int(params['min_child_samples'])

    # Perform n_fold cross validation with hyperparameters
    # Use early stopping and evalute based on ROC AUC
    cv_results = lgb.cv(params, lgb_train, nfold = n_folds, num_boost_round = 10000, 
                        early_stopping_rounds = 100, metrics = 'auc', seed = 50)
  
    # Extract the best score
    best_score = max(cv_results['auc-mean'])
    
    # Loss must be minimized
    loss = 1 - best_score
    
    # Dictionary with information for evaluation
    return {'loss': loss, 'params': params, 'status': STATUS_OK}

space = {
#    'class_weight': hp.choice('class_weight', [None, 'balanced']),
#    'boosting_type': 'gbdt',
#    'learning_rate': hp.loguniform('learning_rate', np.log(0.01), np.log(0.2)),
#    'subsample_for_bin': hp.quniform('subsample_for_bin', 20000, 300000, 20000),
#    'num_iterations ' : hp.quniform('num_iterations ', 100, 1000, 1),

    'min_child_samples': hp.quniform('min_child_samples', 5, 100, 5), 
    'reg_alpha': hp.uniform('reg_alpha', 0.0, 1.0),
    'reg_lambda': hp.uniform('reg_lambda', 0.0, 1.0),
    'colsample_bytree': hp.uniform('colsample_by_tree', 0.5, 1.0),
    'max_depth' : hp.quniform('max_depth', 3, 10, 1),
    'subsample' : hp.quniform('subsample', 0.6, 1, 0.05),
    'num_leaves': hp.quniform('num_leaves', 20, 150, 1),  
    'bagging_freq': hp.choice('bagging_freq',[1, 2, 4, 8, 10]),
    'objective' : hp.choice('objective',['binary'])
}

trials = Trials()
best = fmin(objective, space, algo=tpe.suggest, trials=trials, max_evals=10)

best['max_depth'] = int(best['max_depth'])
best['num_leaves'] = int(best['num_leaves'])
#best['subsample_for_bin'] = int(best['subsample_for_bin'])
best['min_child_samples'] = int(best['min_child_samples'])
best['objective'] = 'binary'

best_gbm = lgb.train(**best, boosting_type = 'gbdt', learning_rate = 0.05,num_boost_round = 10000, train_set = lgb_train)

The parameters in best are:
{'bagging_freq': 1,
 'colsample_by_tree': 0.6104682509816725,
 'max_depth': 7,
 'min_child_samples': 35,
 'num_leaves': 87,
 'objective': 'binary',
 'reg_alpha': 0.24067677738255566,
 'reg_lambda': 0.7005953473317331,
 'subsample': 0.8}

Could you help me in figuring out what is happening?

Apologies if it turns out to be something stupid. Thank you!

**Comment by StrikerRUS** [comment]
> Why would this be incorrect?

Because this is different APIs. Use scikit-learn modules for feature selection if you want to use LightGBM sklearn wrapper.

> The error is : train() got an unexpected keyword argument 'bagging_freq'. 

Fix this line:

```
best_gbm = lgb.train(params=best, boosting_type = 'gbdt', learning_rate = 0.05,num_boost_round = 10000, train_set = lgb_train)
```

**Comment by Sanchita-P** [comment]
> Because this is different APIs. Use scikit-learn modules for feature selection if you want to use LightGBM sklearn wrapper.

Do you mean to use scikit-learn modules for hyperparameter tuning in the cross-validation module? If they are different APIs, what is the risk of happening?

**Comment by StrikerRUS** [comment]
I mean, scikit-learn modules for feature selection: https://scikit-learn.org/stable/modules/classes.html#module-sklearn.feature_selection, if you want to use `lgb.LGBMClassifier`/`lgb.LGBMRegressor`/`lgb.LGBMRanker`.

>  If they are different APIs, what is the risk of happening?

The are incompatible. The are not designed to be used together. There is no warranty that they can be used together. You can search this repo for errors people faced when combined standard and sklearn APIs.

**Comment by Sanchita-P** [comment]
@StrikerRUS , thanks again for your response! :) I apologize if my questions are naive, I am new to this and hence a steep learning curve. I was earlier using LGBMClassifier, however I also wish to use the cv module of lightGBM, so can no longer use LGBMClassifier of sklearn.

Additionally, earlier you had mentioned that one should use argmax manually to predict a class label. However, predict is returning only 1 value per observation. Is this the probability of the observation being 1?

Thanks again for your help!

**Comment by StrikerRUS** [comment]
> I was earlier using LGBMClassifier, however I also wish to use the cv module of lightGBM, so can no longer use LGBMClassifier of sklearn.

I think that you should continue using `LGBMClassifier` and stick to scikit-learn interface as it's more user-friendly, especially for beginners. In that case, you can replace `lgb.cv` with sciikit-learn cross validation: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate.

 >However, predict is returning only 1 value per observation.

For **binary** problem it returns probabilities for positive class. So yes, no argmax is needed.

**Comment by Sanchita-P** [comment]
All your points have been duly noted, thank you for all the help and patience! Wishing you all the best :)

**Comment by StrikerRUS** [comment]
Your are welcome!
